nnfabrik/templates/transfer/transferred_trained_model.py

```diff
@@ -0,0 +1,310 @@
+                "prev_model_fn",
+                "prev_model_hash",
+                "prev_dataset_fn",
+                "prev_dataset_hash",
+                "prev_trainer_fn",
+                "prev_trainer_hash",
+                "prev_seed",
+                "prev_collapsed_history",
+            )
+
+            # get the current transfer step
+            max_transfer_step = prev_transferred_model.fetch("transfer_step").max() if prev_transferred_model else 0
+
+            for transfer_step in range(1, max_transfer_step + 1):
+                # get the transfer recipe
+                recipe = self._transfer_recipe(transfer_step)
+                post_restr = recipe.post_restr if recipe else dj.AndList([])
+
+                # apply the recipe
+                transfer_from = prev_transferred_model * recipe
+                transfers = (
+                    self.model_table * self.dataset_table * self.trainer_table * self.seed_table * transfer_from
+                )  # combine recipe restriction with all possible training combinations
+
+                transfers = transfers * self.CollapsedHistory().proj(
+                    "collapsed_history",
+                    "prev_collapsed_history",
+                    prev_model_fn="model_fn",
+                    prev_model_hash="model_hash",
+                    prev_dataset_fn="dataset_fn",
+                    prev_dataset_hash="dataset_hash",
+                    prev_trainer_fn="trainer_fn",
+                    prev_trainer_hash="trainer_hash",
+                    prev_seed="seed",
+                )  # map previous transferred model to its collapsed history
+
+                transfers = (
+                    dj.U(
+                        "transfer_step",
+                        "model_fn",
+                        "model_hash",
+                        "dataset_fn",
+                        "dataset_hash",
+                        "trainer_fn",
+                        "trainer_hash",
+                        "seed",
+                        "collapsed_history",
+                        "data_transfer",
+                    )
+                    & transfers
+                    & post_restr  # restrict with post_rest
+                )
+
+                if key_source is not None:
+                    key_source = key_source + transfers.proj()
+                else:
+                    key_source = transfers.proj()
+
+        # normal entries as a combination of Dataset, Model, Trainer, and Seed tables
+        step_0 = self.model_table * self.dataset_table * self.trainer_table * self.seed_table
+        # add transfer_step, collapsed_history and data_transfer as prim keys
+        base = dj.U("transfer_step", "collapsed_history", "data_transfer",) * step_0.proj(
+            transfer_step="0",
+            collapsed_history='""',
+            data_transfer="0",
+        )
+
+        if key_source is None:
```

Comment:
I suggest you create `base` at the very beginning of the function and set `key_source = base` first. And then just keep on adding `transfers.proj()` to it in the for loop for `transfer` building. You can then skip this whole if/else block here.